lib/curl_multibyte.c

```diff
@@ -88,12 +88,106 @@ char *curlx_convert_wchar_to_UTF8(const wchar_t *str_w)
 
 #if defined(USE_WIN32_LARGE_FILES) || defined(USE_WIN32_SMALL_FILES)
 
+
+/* Fix excessive paths (paths that exceed MAX_PATH length of 260).
+ *
+ * This is a helper function to fix paths that would exceed the MAX_PATH
+ * limitation check done by Windows APIs. It does so by normalizing the passed
+ * in filename or path 'in' to its full canonical path, and if that path is
+ * longer than MAX_PATH then setting 'out' to "\\?\" prefix + that full path.
+ *
+ * For example 'in' filename255chars in current directory C:\foo\bar is
+ * fixed as \\?\C:\foo\bar\filename255chars for 'out' which will tell Windows
+ * it's ok to access that filename even though the actual full path is longer
+ * than 255 chars.
+ *
+ * For non-Unicode builds this function may fail sometimes because only the
+ * Unicode versions of some Windows API functions can access paths longer than
+ * MAX_PATH, for example GetFullPathNameW which is used in this function. When
+ * the full path is then converted from Unicode to multibyte that fails if any
+ * directories in the path contain characters not in the current codepage.
+ */
+static bool fix_excessive_path(const TCHAR *in, TCHAR **out)
+{
+  size_t needed, written;
+  const wchar_t *in_w;
+  wchar_t *fbuf = NULL;
+
+#ifndef _UNICODE
+  wchar_t *ibuf = NULL;
+  char *obuf = NULL;
+#endif
+
+  *out = NULL;
+
+#ifndef _UNICODE
+  /* convert multibyte input to unicode */
+  needed = mbstowcs(NULL, in, 0);
+  if(needed == (size_t)-1 || needed >= (32767 - 4))
+    goto error;
+  ++needed; /* for NUL */
+  ibuf = malloc(needed * sizeof(wchar_t));
+  if(!ibuf)
+    goto error;
+  written = mbstowcs(ibuf, in, needed);
+  if(written == (size_t)-1 || written >= needed)
+    goto error;
+  in_w = ibuf;
+#else
+  in_w = in;
+#endif
+
+  /* get full unicode path of the unicode filename or path */
+  needed = (size_t)GetFullPathNameW(in_w, 0, NULL, NULL);
+  if(!needed || needed > (32767 - 4))
+    goto error;
+  /* ignore paths which are not excessive and don't need modification */
+  if(needed <= MAX_PATH)
+    goto error;
+  fbuf = malloc((needed + 4)* sizeof(wchar_t));
+  if(!fbuf)
+    goto error;
+  wcsncpy(fbuf, L"\\\\?\\", 4);
+  written = (size_t)GetFullPathNameW(in_w, needed, fbuf + 4, NULL);
+  if(!written || written >= needed)
+    goto error;
+
+#ifndef _UNICODE
+  /* convert unicode full path to multibyte output */
+  needed = wcstombs(NULL, fbuf, 0);
+  if(needed == (size_t)-1 || needed >= 32767)
+    goto error;
+  ++needed; /* for NUL */
+  obuf = malloc(needed);
+  if(!obuf)
+    goto error;
+  written = wcstombs(obuf, fbuf, needed);
+  if(written == (size_t)-1 || written >= needed)
+    goto error;
+  *out = obuf;
+  obuf = NULL;
+#else
+  *out = fbuf;
+  fbuf = NULL;
+#endif
+
+error:
```

Comment:
Perhaps use `cleanup:` instead of `error:`, since cases like this are not actually errors yet it says `goto error`:
```c
/* ignore paths which are not excessive and don't need modification */
```

Reply:
Anything that makes the function fail is an error so I'm going to leave it. When we skip paths, the function fails (returns false).